vault/quotas/quotas_rate_limit_test.go

```diff
@@ -116,3 +119,79 @@ func TestRateLimitQuota_Allow(t *testing.T) {
 		require.Falsef(t, numAllow > want, "too many successful requests; addr: %s, want: %d, numSuccess: %d, numFail: %d, elapsed: %d", addr, want, numAllow, numFail, elapsed)
 	}
 }
+
+func TestRateLimitQuota_Allow_WithBlock(t *testing.T) {
+	rlq := &RateLimitQuota{
+		Name:          "test-rate-limiter",
+		Type:          TypeRateLimit,
+		NamespacePath: "qa",
+		MountPath:     "/foo/bar",
+		Rate:          16.7,
+		BlockInterval: 10 * time.Second,
+
+		// override values to lower durations for testing purposes
+		purgeInterval: 10 * time.Second,
+		staleAge:      10 * time.Second,
+	}
+
+	require.NoError(t, rlq.initialize(logging.NewVaultLogger(log.Trace), metricsutil.BlackholeSink()))
+	require.True(t, rlq.getPurgeBlocked())
+
+	var wg sync.WaitGroup
+
+	reqFunc := func(addr string, atomicNumAllow, atomicNumFail *atomic.Int32) {
+		defer wg.Done()
+
+		resp, err := rlq.allow(&Request{ClientAddress: addr})
+		if err != nil {
+			return
+		}
+
+		if resp.Allowed {
+			atomicNumAllow.Add(1)
+		} else {
+			atomicNumFail.Add(1)
+		}
+	}
+
+	results := make(map[string]*clientResult)
+
+	start := time.Now()
+	end := start.Add(5 * time.Second)
+
+	for time.Now().Before(end) {
+		for i := 0; i < 5; i++ {
+			wg.Add(1)
+
+			addr := fmt.Sprintf("127.0.0.%d", i)
+			cr, ok := results[addr]
+			if !ok {
+				results[addr] = &clientResult{atomicNumAllow: atomic.NewInt32(0), atomicNumFail: atomic.NewInt32(0)}
+				cr = results[addr]
+			}
+
+			go reqFunc(addr, cr.atomicNumAllow, cr.atomicNumFail)
+
+			time.Sleep(2 * time.Millisecond)
+		}
+	}
+
+	wg.Wait()
+
+	for _, cr := range results {
+		numAllow := cr.atomicNumAllow.Load()
+		numFail := cr.atomicNumFail.Load()
+
+		// Since blocking is enabled, each client should only have 'rate' successful
+		// requests, whereas all subsequent requests fail.
+		require.Equal(t, int32(17), numAllow)
+		require.NotZero(t, numFail)
+	}
+
+	// allow enough time for all blocked clients to be purged
+	time.Sleep(rlq.purgeInterval * 2)
```

Comment:
I wonder if there's some way to speed this up so the test does not take 20 seconds.

I wrote some code to fake time.Now() and tickers in the gauge collection library, and Nick suggested I move that into a common library for re-use.

Not a blocker, though.

Reply:
We just need to wait for the garbage collector to run. A multiple of two was added here for safety, but we can omit that entirely I think as just rely on a ticker that checks every second or timesout at `purgeInterval`. I've updated the test to reflect this. If you think there is a cleaner way, please lmk 👍